File: Algorithims and Data Structure II/P2. Força Bruta, Backtracking - Método Guloso/Backtracking e Método Guloso - Trabalho/1687.py
def mdc(a: int, b: int) -> int:
    while b !=0:
        resto = a % b
        a = b
        b = resto

    return a

def verifica(xi: int, yi: int, xf: int, yf: int) -> bool:
    dist_linhas = abs(xi - xf)
    dist_colunas = abs(yi - yf)
    
    return mdc(dist_linhas, dist_colunas) == 1

def backtrack(x: int, y: int, np: int, n: int, p: int) -> int:
    global cont_solucao

    if (np == p):
        cont_solucao += 1
        return
    
    for i in range(n):
        for j in range(n):
            if not verifica(x, y, i, j):
                continue
            backtrack(i, j, np + 1, n, p)

def solucao(n: int, p: int) -> int:
    for i in range(n):
        for j in range(n):
            backtrack(i, j, 1, n, p)
    
    return cont_solucao

while True:
    n, p = map(int, input().split())
    if (n == 0) and (p == 0):
        break
    cont_solucao = 0
    print(solucao(n, p) % 1300031)

File: Algorithims and Data Structure II/P2. Força Bruta, Backtracking - Método Guloso/Backtracking e Método Guloso - Trabalho/1687.c
#include <stdio.h>

int cont_solucao = 0;

int mdc(int a, int b) {
    int resto;

    while (b != 0) {
        resto = a % b;
        a = b;
        b = resto;
    }

    return a;
}

int abs(int a) {
    return (a > 0) ? (a) : (a * (-1));
}

int verifica(int xi, int yi, int xf, int yf) {
    int dl = abs(xi - xf);
    int dc = abs(yi - yf);

    return mdc(dl, dc) == 1;
}

void backtracking(int n, int p, int x, int y, int np) {
    if (np == p) {
        cont_solucao++;
        return;
    }

    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            if (verifica(x, y, i, j))
                backtracking(n, p, i, j, np + 1);
}

void solucao(int n, int p) {
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            backtracking(n, p, i, j, 1);
}

int main() {
    int n, p;

    while (1) {
        scanf("%d %d", &n, &p);
        if (!n && !p)
            break;
        solucao(n, p);
        printf("%d\n", cont_solucao % 1300031);
        cont_solucao = 0;
    }

    return 0;
}